File: CSES/Dynamic Programming/q12_Rectangle_Cutting/bu.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

int main() {
    int a, b;
    cin >> a >> b;

    // dp[i][j] = minimum number of cuts for i x j rectangle
    vector<vector<int>> dp(a + 1, vector<int>(b + 1, 0));

    for (int i = 1; i <= a; ++i) {
        for (int j = 1; j <= b; ++j) {
            if (i == j) {
                dp[i][j] = 0; // already a square
            } else {
                int minCuts = INT_MAX;

                // vertical cuts (cut along height)
                for (int k = 1; k < i; ++k) {
                    minCuts = min(minCuts, 1 + dp[k][j] + dp[i - k][j]);
                }

                // horizontal cuts (cut along width)
                for (int k = 1; k < j; ++k) {
                    minCuts = min(minCuts, 1 + dp[i][k] + dp[i][j - k]);
                }

                dp[i][j] = minCuts;
            }
        }
    }

    cout << dp[a][b] << endl;
    return 0;
}

File: CSES/Dynamic Programming/q12_Rectangle_Cutting/bu.py
a, b = map(int, input().split())

# dp[i][j] = min number of cuts to divide i x j rectangle into squares
dp = [[0] * (b + 1) for _ in range(a + 1)]

for i in range(1, a + 1):
    for j in range(1, b + 1):
        if i == j:
            dp[i][j] = 0  # already a square, no cut needed
        else:
            min_cuts = float("inf")
            # try all vertical cuts
            for k in range(1, i):
                min_cuts = min(min_cuts, 1 + dp[k][j] + dp[i - k][j])
            # try all horizontal cuts
            for k in range(1, j):
                min_cuts = min(min_cuts, 1 + dp[i][k] + dp[i][j - k])
            dp[i][j] = min_cuts

print(dp[a][b])

# Python TLE even here. Java and cpp accepted.
